System.DllNotFoundException: Std-symbols (Mono)

##### Brief Description

Std-Symbols missing when attempting to call generated function
`// void LoadClientScript(std::string name, char* buffer, size_t size)
void LoadClientScript(string name, sbyte* buffer, ulong size);`

OS: Windows

##### Used settings

Using CppSharp NuGet package
Runtime: Mono x64
Target: MSVC

##### Stack trace or incompilable generated code

`System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.DllNotFoundException: Std-symbols
  at (wrapper managed-to-native) Std.Allocator+__Internal:ctorc__N_std_S_allocator__C_0 (intptr)
  at Std.Allocator..ctor () [0x00040] in <9cd678c433524087a31898f7e73337b9>:0
  at SharpOrange.APIInternal.LoadClientScript (System.String name, System.SByte* buffer, System.UInt64 size) [0x00033] in <9cd678c433524087a31898f7e73337b9>:0
  at SharpOrange.Client.AddScript (System.String file) [0x001d1] in <9cd678c433524087a31898f7e73337b9>:0
  at SharpExample.SharpExample..ctor (SharpOrange.API apip) [0x0005f] in <2d858444b9554777911b21b97c52439c>:0
  at (wrapper managed-to-native) System.Reflection.MonoCMethod:InternalInvoke (System.Reflection.MonoCMethod,object,object[],System.Exception&)
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00002] in <638b7550331b4aebbce60a36c5915ef9>:0
   --- End of inner exception stack trace ---
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00014] in <638b7550331b4aebbce60a36c5915ef9>:0
  at System.Reflection.MonoCMethod.DoInvoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0007a] in <638b7550331b4aebbce60a36c5915ef9>:0
  at System.Reflection.MonoCMethod.Invoke (System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in <638b7550331b4aebbce60a36c5915ef9>:0
  at System.Reflection.ConstructorInfo.Invoke (System.Object[] parameters) [0x00000] in <638b7550331b4aebbce60a36c5915ef9>:0
  at SharpOrange.Server.LoadResource (System.String resource) [0x000d6] in <9cd678c433524087a31898f7e73337b9>:0`

The code generator are missing some }:
![](http://i.imgur.com/6WOpDpF.png)

... so compilation fails. I used to manually add these before using the code.

`C:\Users\Dator\Source\Repos\mono-module\CppSharpOrange\Generated\SharpOrange.cs(1419,10) : error CS1513: } expected
C:\Users\Dator\Source\Repos\mono-module\CppSharpOrange\Generated\SharpOrange.cs(2440,10) : error CS1513: } expected
C:\Users\Dator\Source\Repos\mono-module\CppSharpOrange\Generated\SharpOrange.cs(2457,10) : error CS1513: } expected
C:\Users\Dator\Source\Repos\mono-module\CppSharpOrange\Generated\SharpOrange.cs(2543,10) : error CS1513: } expected
C:\Users\Dator\Source\Repos\mono-module\CppSharpOrange\Generated\SharpOrange.cs(3208,10) : error CS1513: } expected
C:\Users\Dator\Source\Repos\mono-module\CppSharpOrange\Generated\SharpOrange.cs(4229,10) : error CS1513: } expected
C:\Users\Dator\Source\Repos\mono-module\CppSharpOrange\Generated\SharpOrange.cs(4246,10) : error CS1513: } expected
C:\Users\Dator\Source\Repos\mono-module\CppSharpOrange\Generated\SharpOrange.cs(4332,10) : error CS1513: } expected`

--- a/src/Generator/Generators/CSharp/CSharpSources.cs
+++ b/src/Generator/Generators/CSharp/CSharpSources.cs
@@ -1492,6 +1492,7 @@ namespace CppSharp.Generators.CSharp
             }
 
             var marshals = new List<string>();
+            var numBlocks = 0;
             for (int i = 0; i < method.Parameters.Count; i++)
             {
                 var param = method.Parameters[i];
@@ -1515,6 +1516,11 @@ namespace CppSharp.Generators.CSharp
                     Write(marshal.Context.Before);
 
                 marshals.Add(marshal.Context.Return);
+                if (ctx.HasCodeBlock)
+                {
+                    PushIndent();
+                    numBlocks++;
+                }
             }
 
             var hasReturn = !method.OriginalReturnType.Type.IsPrimitiveType(PrimitiveType.Void);
@@ -1568,6 +1574,9 @@ namespace CppSharp.Generators.CSharp
                     WriteLine("return {0};", marshal.Context.Return);
                 }
             }
+
+            for (var i = 0; i < numBlocks; ++i)
+                WriteCloseBraceIndent();
         }
 
         private void InvokeProperty(Declaration method, IEnumerable<string> marshals)

--- a/tests/Common/Common.Tests.cs
+++ b/tests/Common/Common.Tests.cs
@@ -765,4 +765,14 @@ This is a very long string. This is a very long string. This is a very long stri
         var v = Common.FuncUnion(ut);
         Assert.AreEqual(20, v);
     }
+
+    [Test]
+    public void TestVirtualFuncWithStringParams()
+    {
+        using (var VirtFuncWithStringParam = new ImplementsVirtualFunctionsWithStringParams())
+        {
+            VirtFuncWithStringParam.PureVirtualFunctionWithStringParams("anyRandomString");
+            Assert.That(VirtFuncWithStringParam.VirtualFunctionWithStringParam("anyRandomString").Equals(5));
+        }
+    }
 }

--- a/tests/Common/Common.cpp
+++ b/tests/Common/Common.cpp
@@ -756,3 +756,28 @@ int HasCopyAndMoveConstructor::getField()
 {
     return field;
 }
+
+HasVirtualFunctionsWithStringParams::HasVirtualFunctionsWithStringParams()
+{
+}
+
+HasVirtualFunctionsWithStringParams::~HasVirtualFunctionsWithStringParams()
+{
+}
+
+ImplementsVirtualFunctionsWithStringParams::ImplementsVirtualFunctionsWithStringParams()
+{
+}
+
+ImplementsVirtualFunctionsWithStringParams::~ImplementsVirtualFunctionsWithStringParams()
+{
+}
+
+void ImplementsVirtualFunctionsWithStringParams::PureVirtualFunctionWithStringParams(std::string testString)
+{
+}
+
+int HasVirtualFunctionsWithStringParams::VirtualFunctionWithStringParam(std::string testString)
+{
+    return 5;
+}

--- a/tests/Common/Common.h
+++ b/tests/Common/Common.h
@@ -1281,3 +1281,20 @@ public:
 private:
     int field;
 };
+
+class DLL_API HasVirtualFunctionsWithStringParams
+{
+public:
+    HasVirtualFunctionsWithStringParams();
+    ~HasVirtualFunctionsWithStringParams();
+    virtual void PureVirtualFunctionWithStringParams(std::string testString) = 0;
+    virtual int VirtualFunctionWithStringParam(std::string testString);
+};
+
+class DLL_API ImplementsVirtualFunctionsWithStringParams : public HasVirtualFunctionsWithStringParams
+{
+public:
+    ImplementsVirtualFunctionsWithStringParams();
+    ~ImplementsVirtualFunctionsWithStringParams();
+    virtual void PureVirtualFunctionWithStringParams(std::string testString);
+};

